File: code/ch02/2.5.1.threeSumCloset.java
import java.lang.Math;
import java.util.*;

public class Solution {
  public int threeSumClosest(int[] nums, int target) {
    int n = nums.length;
    //if (n < 3) return;
    Arrays.sort(nums);
    int res = nums[0] + nums[1] + nums[2];
    for (int i = 0; i < n - 2; i ++) {
      if (i != 0 && nums[i] == nums[i - 1]) continue;
      int l = i + 1, r = n - 1;
      while (l < r) {
        int s = nums[i] + nums[l] + nums[r];
        if (s == target) return s;
        if (Math.abs(s - target) < Math.abs(res - target))
          res = s;
        if (s < target) l += 1;
        else if (s > target) r -= 1;
      }
    }
    return res;
  }
}

File: code/ch02/2.5.1.threeSumCloset.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        n = len(nums)
        if n < 3:
            return
        nums.sort()
        res = nums[0] + nums[1] + nums[2]
        for i in range(n - 2):
            # 去重
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            # 固定 i，寻找 l 和 r，使用双指针法
            l = i + 1
            r = n - 1
            while l < r:
                s = nums[i] + nums[l] + nums[r]
                if s == target:
                    return s

                if abs(s - target) < abs(res - target):
                    res = s

                if s < target:
                    l += 1
                elif s > target:
                    r -= 1

        return res
